cpp/src/arrow/dataset/file_csv.cc

```diff
@@ -111,9 +111,26 @@ static inline Result<csv::ConvertOptions> GetConvertOptions(
 
   if (!scan_options) return convert_options;
 
-  auto materialized = scan_options->MaterializedFields();
-  std::unordered_set<std::string> materialized_fields(materialized.begin(),
-                                                      materialized.end());
+  auto field_refs = scan_options->MaterializedFields();
+  std::unordered_set<std::string> materialized_fields;
+  materialized_fields.reserve(field_refs.size());
+  // Preprocess field refs. We try to avoid FieldRef::GetFoo here since that's
+  // quadratic (and this is significant overhead with 1000+ columns)
+  for (const auto& ref : field_refs) {
+    if (const auto* name = ref.name()) {
```

Comment:
Nit: Maybe it's just me but I have to look up `const auto*` every time to remember what the rules are since we so rarely return raw pointers.  Maybe we could just be explicit.
```suggestion
    if (const std::string* name = ref.name()) {
```

Reply:
I'll write these out. I used `const auto*` to stress that we're getting a raw pointer here instead of a reference but it's probably best to be explicit about the types in that case anyways.